usr.sbin/efitable/efitable.c

```diff
@@ -0,0 +1,207 @@
+#include <sys/param.h>
+#include <sys/stat.h>
+#include <err.h>
+#include <fcntl.h>
+#include <getopt.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sysexits.h>
+#include <unistd.h>
+#include <uuid.h>
+
+#define TABLE_MAX_LEN 30
+
+#ifdef EFI_TABLE_ESRT
+#undef EFI_TABLE_ESRT
+#endif
+
+#define	EFI_TABLE_ESRT					\
+			{0xb122a263,0x3661,0x4f68,0x99,0x29,{0x78,0xf8,0xb0,0xd6,0x21,0x80}}
+
+static void efi_table_print_esrt(const void *data);
+static void usage(void);
+static void fail(const char msg[]);
+
+struct efi_table_op {
+	char name[TABLE_MAX_LEN];
+	void (*parse) (const void *);
+	struct uuid uuid;
+};
+
+static const struct efi_table_op efi_table_ops[] = {
+	{ .name = "esrt",
+		.parse = efi_table_print_esrt,
+		.uuid = EFI_TABLE_ESRT },
+};
+
+int
+main(int argc, char **argv)
+{
+	struct efi_get_table_ioc table = {
+		.ptr = NULL,
+		.buf_len = 0,
+		.table_len = 0
+	};
+	int efi_fd, ch, rc = 1, efi_idx = -1;
+	bool got_table = false;
+	bool table_set = false;
+	bool uuid_set = false;
+	struct option longopts[] = {
+		{ "uuid",  required_argument, NULL, 'u' },
+		{ "table", required_argument, NULL, 't' },
+		{ NULL,    0,                 NULL,  0  }
+	};
+
+	while ((ch = getopt_long(argc, argv, "u:t:", longopts, NULL)) != -1) {
+		switch (ch) {
+			case 'u': {
+					  char *uuid_str = optarg;
+					  struct uuid uuid;
+					  uint32_t status;
+					  size_t n;
+
+					  uuid_set = 1;
+
+					  uuid_from_string(uuid_str, &uuid, &status);
+					  if (status != uuid_s_ok)
+						  fail("invalid UUID");
+
+					  for (n = 0; n < nitems(efi_table_ops); n++) {
+						  if (!memcmp(&uuid,
+									  &efi_table_ops[n].uuid,
+									  sizeof(uuid))) {
+							  efi_idx = n;
+							  got_table = true;
+							  break;
+						  }
+					  }
+					  break;
+				  }
+			case 't': {
+					  char *table_name = optarg;
+					  size_t n;
+
+					  table_set = true;
+
+					  for (n = 0; n < nitems(efi_table_ops); n++) {
+						  if (!strcmp(table_name,
+									  efi_table_ops[n].name)) {
+							  efi_idx = n;
+							  got_table = true;
+							  break;
+						  }
+					  }
+
+					  if (!got_table)
+						  fail("unsupported efi table");
+					  break;
+				  }
+			default:
+				  usage();
+		}
+	}
+
+	if (!table_set && !uuid_set)
+		fail("table is not set");
+
+	if (!got_table)
+		fail("unsupported table");
+
+	efi_fd = open("/dev/efi", O_RDWR);
+	if (efi_fd < 0)
+		fail("/dev/efi");
+
+	table.uuid = efi_table_ops[efi_idx].uuid;
+	if (ioctl(efi_fd, EFIIOC_GET_TABLE, &table) == -1)
+		fail(NULL);
+
+	table.ptr = malloc(table.table_len);
+	table.buf_len = table.table_len;
+
+	if (ioctl(efi_fd, EFIIOC_GET_TABLE, &table) == -1)
+		fail(NULL);
+	close(efi_fd);
+
+	efi_table_ops[efi_idx].parse(table.ptr);
+	return (rc);
+}
+
+static void
+efi_table_print_esrt(const void *data)
+{
+	const struct efi_esrt_table *esrt = data;
+	const struct efi_esrt_entry_v1 *esre = (void *)esrt->entries;
```

Comment:
@SergiiDmytruk can you look at this? I have `Segmentation fault (core dumped)` here

Reply:
After the first `ioctl` call `table` changes from:
```
{uuid = {time_low = 2971837027, time_mid = 13921, time_hi_and_version = 20328, clock_seq_hi_and_reserved = 153 '\231', clock_seq_low = 41 ')', node = "x\370\260\326!\200"},
 ptr = 0x0,
 table_len = 0,
 buf_len = 0}
```
to
```
{uuid = {time_low = 2971837027, time_mid = 13921, time_hi_and_version = 20328, clock_seq_hi_and_reserved = 153 '\231', clock_seq_low = 41 ')', node = "x\370\260\326!\200"},
 ptr = 0xffffb8007f942898,
 table_len = 0,
 buf_len = 0}
```
Then `malloc(0)` is called and you can only `free()` its result, there is nothing to access. The kernel should set `.table_set` on first call and copy data into buffer on second call. It does something weird by setting `.ptr` which it should only read.

Reply:
@SergiiDmytruk Ok, I see that table_len and buf_len are not used, probably because I miss a similar part of code: https://github.com/3mdeb/openbsd-src/compare/8b50006...3mdeb:openbsd-src:esrt-api#diff-33b0790f7b42bea48c12fff0d6bed9132dff4ccf41e8ac581446422b12aafe8dR398

I will try to implement something similar in DragonflyBSD